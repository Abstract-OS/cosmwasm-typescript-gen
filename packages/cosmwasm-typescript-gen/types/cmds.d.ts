export namespace Commands {
    export { _boilerplate as boilerplate };
    export { _generate as generate };
    export { _recoil as recoil };
}
import _boilerplate from "./commands/boilerplate";
import _generate from "./commands/generate";
import _recoil from "./commands/recoil";
