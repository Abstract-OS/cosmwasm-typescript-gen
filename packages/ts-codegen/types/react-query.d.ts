declare const _default: (name: string, schemas: any[], outPath: string, options?: { optionalClient?: boolean }) => Promise<void>;
export default _default;
