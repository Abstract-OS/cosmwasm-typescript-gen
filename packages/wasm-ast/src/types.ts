export declare type fieldType = 'Long' | 'Coin' | 'Duration' | 'Height' | string;

export interface Field {
    name: string;
    type: fieldType;
    node: any;
};
export interface Interface {
    name: string;
    fields: Field[]
};
