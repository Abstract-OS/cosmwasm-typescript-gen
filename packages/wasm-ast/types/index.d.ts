export * from './types';
export * from './amino';
export * from './messages';
export * from './utils';
