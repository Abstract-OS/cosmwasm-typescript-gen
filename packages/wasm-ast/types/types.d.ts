export declare type fieldType = 'Long' | 'Coin' | 'Duration' | 'Height' | string;
export interface Field {
    name: string;
    type: fieldType;
    node: any;
}
export interface Interface {
    name: string;
    fields: Field[];
}
export interface MessageSchema extends Interface {
    typeUrl: string;
    name: string;
    fields: Field[];
}
export interface EnumConverter {
    funcName: string;
    convertType: 'to' | 'from';
    type: string;
}
export interface Enum {
    filename?: string;
    name: string;
    node: any;
    from?: EnumConverter;
    to?: EnumConverter;
}
export interface Mutation {
    typeUrl: string;
    TypeName: string;
    methodName: string;
}
