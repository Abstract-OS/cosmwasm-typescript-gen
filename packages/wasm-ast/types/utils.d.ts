import * as t from '@babel/types';
import { Field } from './types';
declare const BILLION: t.NumericLiteral;
export { BILLION };
export declare const recursiveNamespace: (names: any, moduleBlockBody: any) => any;
export declare const arrayTypeNDimensions: (body: any, n: any) => any;
export declare const FieldTypeAsts: {
    string: () => t.TSStringKeyword;
    array: (type: any) => t.TSArrayType;
    Duration: () => t.TSTypeReference;
    Height: () => t.TSTypeReference;
    Coin: () => t.TSTypeReference;
    Long: () => t.TSTypeReference;
};
export declare const typeUrlToAmino: (str: any, MsgName: any) => any;
export declare const shorthandProperty: (prop: string) => t.ObjectProperty;
export declare const importStmt: (names: string[], path: string) => t.ImportDeclaration;
export declare const importAminoMsg: () => t.ImportDeclaration;
export declare const getFieldDimensionality: (field: Field) => {
    typeName: string;
    dimensions: number;
    isArray: boolean;
};
export declare const memberExpressionOrIdentifier: (names: any) => any;
export declare const memberExpressionOrIdentifierSnake: (names: any) => any;
