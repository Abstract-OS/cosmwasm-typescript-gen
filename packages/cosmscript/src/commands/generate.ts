import { prompt } from '../prompt';
import cosmscript from '../index';
import { sync as glob } from 'glob';
import { readFileSync } from 'fs';

export default async (argv) => {

    const questions = [
        {
            _: true,
            type: 'path',
            name: 'jsonSchemaPath',
            message: 'which directory contains the the Rust contracts?',
            default: './rust'
        },
        {
            _: true,
            type: 'path',
            name: 'outPath',
            message: 'where is the output directory?',
            default: './src'
        },
        {
            _: true,
            type: 'string',
            name: 'name',
            message: 'contract name?'
        }
    ];

    const { jsonSchemaPath, outPath, name } = await prompt(questions, argv);

    const files = glob(jsonSchemaPath + '/**/*.json');
    const schemas = files.map(file => JSON.parse(readFileSync(file, 'utf-8')));
    cosmscript(name, schemas, outPath);
};