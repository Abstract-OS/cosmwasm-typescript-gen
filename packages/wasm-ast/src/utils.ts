import * as t from '@babel/types';
import { snake } from "case";
import { Field } from './types';

export const recursiveNamespace = (names, moduleBlockBody) => {
    if (!names || !names.length) return moduleBlockBody;
    const name = names.pop();
    const body = [
        t.exportNamedDeclaration(
            t.tsModuleDeclaration(
                t.identifier(name),
                t.tsModuleBlock(recursiveNamespace(names, moduleBlockBody))
            )
        )
    ];
    return body;
};

export const arrayTypeNDimensions = (body, n) => {
    if (!n) return t.tsArrayType(body);
    return t.tsArrayType(
        arrayTypeNDimensions(body, n - 1)
    );
};

export const FieldTypeAsts = {
    string: () => {
        return t.tsStringKeyword()
    },
    array: (type) => {
        return t.tsArrayType(FieldTypeAsts[type]())
    },
    Duration: () => {
        return t.tsTypeReference(t.identifier('Duration'))
    },
    Height: () => {
        return t.tsTypeReference(t.identifier('Height'))
    },
    Coin: () => {
        return t.tsTypeReference(t.identifier('Coin'))
    },
    Long: () => {
        return t.tsTypeReference(t.identifier('Long'))
    }
};

export const shorthandProperty = (prop: string) => {
    return t.objectProperty(t.identifier(prop), t.identifier(prop), false, true);
};

export const importStmt = (names: string[], path: string) => {
    return t.importDeclaration(
        names.map(name => t.importSpecifier(t.identifier(name), t.identifier(name))),
        t.stringLiteral(path));
};

export const importAminoMsg = () => {
    return importStmt(['AminoMsg'], '@cosmjs/amino');
};

export const getFieldDimensionality = (field: Field) => {
    let typeName = field.type;
    const isArray = typeName.endsWith('[]');
    let dimensions = 0;
    if (isArray) {
        dimensions = typeName.match(/\[\]/g).length - 1;
        typeName = typeName.replace(/\[\]/g, '');
    }
    return {
        typeName,
        dimensions,
        isArray
    };
}

export const memberExpressionOrIdentifier = (names) => {
    if (names.length === 1) {
        return t.identifier(names[0])
    }
    if (names.length === 2) {
        const [b, a] = names;
        return t.memberExpression(
            t.identifier(a),
            t.identifier(b)
        );
    }
    const [name, ...rest] = names;

    return t.memberExpression(
        memberExpressionOrIdentifier(rest),
        t.identifier(name)
    )
};

export const memberExpressionOrIdentifierSnake = (names) => {
    if (names.length === 1) {
        return t.identifier(snake(names[0]))
    }
    if (names.length === 2) {
        const [b, a] = names;
        return t.memberExpression(
            t.identifier(snake(a)),
            t.identifier(snake(b))
        );
    }
    const [name, ...rest] = names;

    return t.memberExpression(
        memberExpressionOrIdentifierSnake(rest),
        t.identifier(snake(name))
    )
};
