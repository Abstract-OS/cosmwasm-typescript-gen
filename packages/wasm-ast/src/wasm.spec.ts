import { importStmt } from './utils'
import generate from '@babel/generator';
import * as t from '@babel/types';

import query_msg from './__fixtures__/schema/query_msg.json';
import execute_msg from './__fixtures__/schema/execute_msg_for__empty.json';

import {
  readonlyClass
} from './wasm'

const expectCode = (ast) => {
  expect(
    generate(ast).code
  ).toMatchSnapshot();
}

const printCode = (ast) => {
  console.log(
    generate(ast).code
  );
}

it('readonly classes', () => {
  printCode(readonlyClass(
    'SG721QueryClient',
    'SG721ReadOnlyInstance',
    query_msg
  ))
});
