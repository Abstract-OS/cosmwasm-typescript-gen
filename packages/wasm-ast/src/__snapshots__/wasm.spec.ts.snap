// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute classes 1`] = `
"export class SG721Client extends SG721QueryClient implements SG721Instance {
  client: SigningCosmWasmClient;
  sender: string;
  contractAddress: string;

  constructor(client: SigningCosmWasmClient, sender: string, contractAddress: string) {
    super(client, contractAddress);
    this.client = client;
    this.contractAddress = contractAddress;
    this.transferNft = this.transferNft.bind(this);
    this.sendNft = this.sendNft.bind(this);
    this.approve = this.approve.bind(this);
    this.revoke = this.revoke.bind(this);
    this.approveAll = this.approveAll.bind(this);
    this.revokeAll = this.revokeAll.bind(this);
    this.mint = this.mint.bind(this);
    this.burn = this.burn.bind(this);
  }

  transferNft = async (recipient?: string, tokenId?: string): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      transfer_nft: {
        recipient,
        token_id: tokenId
      }
    }, \\"auto\\");
  };
  sendNft = async (contract?: string, msg?: BinaryType, tokenId?: string): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      send_nft: {
        contract,
        msg,
        token_id: tokenId
      }
    }, \\"auto\\");
  };
  approve = async (expires?: Expiration, spender?: string, tokenId?: string): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      approve: {
        expires,
        spender,
        token_id: tokenId
      }
    }, \\"auto\\");
  };
  revoke = async (spender?: string, tokenId?: string): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      revoke: {
        spender,
        token_id: tokenId
      }
    }, \\"auto\\");
  };
  approveAll = async (expires?: Expiration, operator?: string): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      approve_all: {
        expires,
        operator
      }
    }, \\"auto\\");
  };
  revokeAll = async (operator?: string): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      revoke_all: {
        operator
      }
    }, \\"auto\\");
  };
  mint = async (): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      mint: {}
    }, \\"auto\\");
  };
  burn = async (tokenId?: string): Promise<ExecuteResult> => {
    return await this.client.execute(this.sender, this.contractAddress, {
      burn: {
        token_id: tokenId
      }
    }, \\"auto\\");
  };
}"
`;

exports[`query classes 1`] = `
"export class SG721QueryClient implements SG721ReadOnlyInstance {
  client: CosmWasmClient;
  contractAddress: string;

  constructor(client: CosmWasmClient, contractAddress: string) {
    this.client = client;
    this.contractAddress = contractAddress;
    this.ownerOf = this.ownerOf.bind(this);
    this.approval = this.approval.bind(this);
    this.approvals = this.approvals.bind(this);
    this.allOperators = this.allOperators.bind(this);
    this.numTokens = this.numTokens.bind(this);
    this.contractInfo = this.contractInfo.bind(this);
    this.nftInfo = this.nftInfo.bind(this);
    this.allNftInfo = this.allNftInfo.bind(this);
    this.tokens = this.tokens.bind(this);
    this.allTokens = this.allTokens.bind(this);
    this.minter = this.minter.bind(this);
    this.collectionInfo = this.collectionInfo.bind(this);
  }

  ownerOf = async (includeExpired?: boolean, tokenId?: string): Promise<OwnerOfResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      owner_of: {
        include_expired: includeExpired,
        token_id: tokenId
      }
    });
  };
  approval = async (includeExpired?: boolean, spender?: string, tokenId?: string): Promise<ApprovalResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      approval: {
        include_expired: includeExpired,
        spender,
        token_id: tokenId
      }
    });
  };
  approvals = async (includeExpired?: boolean, tokenId?: string): Promise<ApprovalsResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      approvals: {
        include_expired: includeExpired,
        token_id: tokenId
      }
    });
  };
  allOperators = async (includeExpired?: boolean, limit?: number, owner?: string, startAfter?: string): Promise<AllOperatorsResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      all_operators: {
        include_expired: includeExpired,
        limit,
        owner,
        start_after: startAfter
      }
    });
  };
  numTokens = async (): Promise<NumTokensResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      num_tokens: {}
    });
  };
  contractInfo = async (): Promise<ContractInfoResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      contract_info: {}
    });
  };
  nftInfo = async (tokenId?: string): Promise<NftInfoResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      nft_info: {
        token_id: tokenId
      }
    });
  };
  allNftInfo = async (includeExpired?: boolean, tokenId?: string): Promise<AllNftInfoResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      all_nft_info: {
        include_expired: includeExpired,
        token_id: tokenId
      }
    });
  };
  tokens = async (limit?: number, owner?: string, startAfter?: string): Promise<TokensResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      tokens: {
        limit,
        owner,
        start_after: startAfter
      }
    });
  };
  allTokens = async (limit?: number, startAfter?: string): Promise<AllTokensResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      all_tokens: {
        limit,
        start_after: startAfter
      }
    });
  };
  minter = async (): Promise<MinterResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      minter: {}
    });
  };
  collectionInfo = async (): Promise<CollectionInfoResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      collection_info: {}
    });
  };
}"
`;

exports[`query interfaces 1`] = `
"export interface SG721ReadOnlyInstance {
  contractAddress: string;
  ownerOf: (includeExpired?: boolean, tokenId?: string) => Promise<OwnerOfResponse>;
  approval: (includeExpired?: boolean, spender?: string, tokenId?: string) => Promise<ApprovalResponse>;
  approvals: (includeExpired?: boolean, tokenId?: string) => Promise<ApprovalsResponse>;
  allOperators: (includeExpired?: boolean, limit?: number, owner?: string, startAfter?: string) => Promise<AllOperatorsResponse>;
  numTokens: () => Promise<NumTokensResponse>;
  contractInfo: () => Promise<ContractInfoResponse>;
  nftInfo: (tokenId?: string) => Promise<NftInfoResponse>;
  allNftInfo: (includeExpired?: boolean, tokenId?: string) => Promise<AllNftInfoResponse>;
  tokens: (limit?: number, owner?: string, startAfter?: string) => Promise<TokensResponse>;
  allTokens: (limit?: number, startAfter?: string) => Promise<AllTokensResponse>;
  minter: () => Promise<MinterResponse>;
  collectionInfo: () => Promise<CollectionInfoResponse>;
}"
`;
